model/jpa/src/main/java/org/keycloak/models/jpa/RoleAdapter.java

```diff
@@ -89,30 +96,88 @@ public void setName(String name) {
 
     @Override
     public boolean isComposite() {
-        return getCompositesStream().count() > 0;
+        TypedQuery<String> query = em.createNamedQuery("getChildrenRoleIds", String.class);
+        query.setParameter("roleId", getId());
+        query.setMaxResults(1);
+        return !query.getResultList().isEmpty();
     }
 
     @Override
     public void addCompositeRole(RoleModel role) {
+        // Avoid lazy loading the composite role collection if not already done
+        // Not using Persistence.getPersistenceUtil().isLoaded(Object, String) as this is not working
+        // properly (Hibernate 5.6) - still returning false even after the collection was lazy loaded
+        if (Hibernate.isPropertyInitialized(getEntity(), "compositeRoles")) {
+            addCompositeRoleUsingLoadedCompositeCollection(role);
+        }
+        else {
+            addCompositeRoleWithoutLoadingCompositeCollection(role);
+        }
```

Comment:
I ran the test `CreateRealms` test case that uses the Admin-CLI-Token.

When building and remote-debugging the legacy Wildfly distribution, `compositeRoles` was always initialized. 

Doing the same for the Quarkus distribution, the code was hitting the other path were the properties weren't initialized correctly. 

Given the situation that the legacy distribution runs as before, and the Quarkus distribution is now faster, this would be ok to me. Still it is quite annoying, as when debugging "as usual" with the KeycloakServer in test-utils (that is using Undertow), it shows the same effects as the Wildfly this can't be debugged nicely. It seems that whenever the Role is fetched by ID, it doesn't apply any lazy loading.

Why thoughts on this?

Reply:
> When building and remote-debugging the legacy Wildfly distribution, compositeRoles was always initialized.
Doing the same for the Quarkus distribution, the code was hitting the other path were the properties weren't initialized correctly.

Exactly, this is what I mentioned above - this was the reason why tests were failing with Quarkus while still running fine with Undertow and Wildfly. It took me a while to figure this out: Hibernate lazy-loading behavior is different with Quarkus than with Wildfly/Undertow.
And this is not a Hibernate version difference (upgrading Hibernate to match Quarkus' gives to the same results).

Hibernate behavior with Quarkus is the most optimal one - meaning lazy loading properly working.


> Given the situation that the legacy distribution runs as before, and the Quarkus distribution is now faster, this would be ok to me.

I agree, but I'm annoyed having the same code base providing two different behaviors depending on the container it's deployed in...

> Still it is quite annoying, as when debugging "as usual" with the KeycloakServer in test-utils (that is using Undertow), it shows the same effects as the Wildfly this can't be debugged nicely. It seems that whenever the Role is fetched by ID, it doesn't apply any lazy loading.
Why thoughts on this?

Well not really, apart that I suspect a difference in Hibernate settings or in bytecode augmentation (this one is for sure because of Quarkus), but I did not dig into this deeper.

To investigate this, I would start from main again, add only the `Hibernate.isPropertyInitialized(getEntity(), "compositeRoles")` line (log + a boolean) to check whether this is related to this PR (but I doubt it - I think this is an older problem).

Reply:
@davoustp - thanks, let's keep it as is, then. 

No-one will want to change an existing behavior in Wildfly/Untertow. I suspect it would affect other places as well, and that could have unknown positive or negative impact on performance. 

Again, let's not touch that. Consider my comment to be resolved.